tensorflow/python/keras/layers/normalization.py

```diff
@@ -613,8 +613,50 @@ def _fake_update():
 
     return (r, d, out_mean, out_variance)
 
-  def _moments(self, inputs, reduction_axes, keep_dims):
-    mean, variance = nn.moments(inputs, reduction_axes, keep_dims=keep_dims)
+  def moment_calculation(self,
+                         inputs,
+                         axes,
+                         keep_dims=False,
+                         mask=None):
+
+      with ops.name_scope("moments", values=[inputs, axes]):
+          # The dynamic range of fp16 is too limited to support the collection of
+          # sufficient statistics. As a workaround we simply perform the operations
+          # on 32-bit floats before converting the mean and variance back to fp16
+          casted_inputs = math_ops.cast(inputs, dtypes.float32) if inputs.dtype == dtypes.float16 else inputs
```

Comment:
i am curious about the fp16 issue here, since it is kind of orthogonal with regard to masking. I think we should separate this into another change.

Reply:
This and a lot of the previous changes came directly from tf.nn.moments here:
https://github.com/tensorflow/tensorflow/blob/5e2a91f65cfb23f996136b9201d9312c9c36b941/tensorflow/python/ops/nn_impl.py#L1023
Would it make more sense to just modify this function to take a mask, instead of duplicating the code? It seems like we're creating a lot of extra stuff to maintain, when the functionality is almost identical.

Reply:
Ah, I see. Thanks for for pointer. I think it actually make more sense to push the logic to tf.nn.moments with the new param of mask.

Reply:
Also adding @alextp who is the general owner of the API since this will touch the tf.nn.moment API to have a new parameter.

Alex, masking is more like a Keras implementation detail for annotating the padded data. I am not very sure if we should push keras logic into lower level ops API.

Reply:
I think an implicit mask in core TF APIs is a no-no. An explicit mask argument though seems useful in general.

Reply:
where do we go from here @qlzh727 ?

Reply:
Passing an explicit mask argument to nn.moments is reasonable to me.